recommenders/models/ncf/dataset.py

```diff
@@ -1,321 +1,450 @@
+                            .union(user_train_data[self.col_item].unique())
+                        user_negative_item_pool = self._get_user_negatives_pool(user_positive_item_pool)
+                        n_samples = self.n_neg_test
+                        n_samples = self._check_sample_size(user, n_samples, user_negative_item_pool, training=False)
+
+                        user_examples_dfs = []
+                        # sample n_neg_test negatives for each positive example and assign a batch index
+                        for positive_example in np.array_split(user_test_data, user_test_data.shape[0]):
+                            negative_examples = self._get_negative_examples(user, user_negative_item_pool, n_samples)
+                            examples = pd.concat([positive_example, negative_examples])
+                            examples[self.col_test_batch] = batch_idx
+                            user_examples_dfs.append(examples)
+                            batch_idx += 1
+
+                        # append user test data to file
+                        user_examples = pd.concat(user_examples_dfs)
+                        user_examples.to_csv(self.test_file_full, mode='a', index=False, header=False)
+
+    def _split_into_batches(self, shuffle_buffer, batch_size):
+        for i in range(0, len(shuffle_buffer), batch_size):
+            yield shuffle_buffer[i:i + batch_size]
+
+    def _prepare_batch_with_id(self, batch):
+        return [
+            [self.user2id[user] for user in batch[self.col_user].values],
+            [self.item2id[item] for item in batch[self.col_item].values],
+            batch[self.col_rating].values.tolist()
         ]
-        df_reindex.columns = [self.col_user, self.col_item, self.col_rating]
 
-        return df_reindex
-
-    def _init_train_data(self):
-        """Return all negative items (in train dataset) and store them in self.interact_status[self.col_item + '_negative']
-        store train dataset in self.users, self.items and self.ratings
+    def _prepare_batch_without_id(self, batch):
+        return [
+            batch[self.col_user].values.tolist(),
+            batch[self.col_item].values.tolist(),
+            batch[self.col_rating].values.tolist()
+        ]
 
+    def _release_shuffle_buffer(self, shuffle_buffer, batch_size, yield_id, write_to=None):
+        prepare_batch = self._prepare_batch_with_id if yield_id else self._prepare_batch_without_id
+        shuffle_buffer_df = pd.concat(shuffle_buffer)
+        shuffle_buffer_df = shuffle_buffer_df.sample(shuffle_buffer_df.shape[0])  # shuffle the buffer
+        for batch in self._split_into_batches(shuffle_buffer_df, batch_size):
+            if batch.shape[0] == batch_size:
+                if write_to:
+                    batch.to_csv(write_to, mode='a', header=False, index=False)
+                yield prepare_batch(batch)
+            else:
+                return batch
```

Comment:
Why is the return statement needed?

Reply:
When the shuffle buffer is released, it yields batches of fixed size. However, there is likely to be a remainder when splitting the shuffle buffer into fixed size batches and so the final return passes this remainder back to the train_loader so that it can start filling the fresh shuffle buffer with this remainder.

Reply:
But train_loader could get that from `yield`, couldn't it? In fact, I think the return syntax will not provide the remainder to the caller of this function, only yield does. Return has no effect in an iterator like this.

Reply:
The `_release_shuffle_buffer` iterator does return batch if it does not have the required number of examples (which can only be the case for the last batch in the shuffle buffer). When the iterator hits the return statement it will stop the execution and return the batch to train_loader. The train_loader uses `yield from` syntax to yield batches from `_release_shuffle_buffer` to the calling function (the NCF model training function) so full batches are simply passed through the train_loader function and only the final incomplete batch is returned to train_loader. If the final incomplete batch were to be yielded, it would immediately be yielded to the model training function which we don't want.

I did try writing the code without this nested iterator approach initially but it was very very messy.